api/v1alpha4/clusterclass_webhooks.go

```diff
@@ -0,0 +1,154 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/google/go-cmp/cmp"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+func (in *ClusterClass) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(in).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1alpha4-clusterclass,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=clusterclasses,versions=v1alpha4,name=validation.clusterclass.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-cluster-x-k8s-io-v1alpha4-clusterclasses,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=cluster.x-k8s.io,resources=clusterclasses,versions=v1alpha4,name=default.clusterclass.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+var _ webhook.Validator = &ClusterClass{}
+var _ webhook.Defaulter = &ClusterClass{}
+
+// Default satisfies the defaulting webhook interface.
+func (in *ClusterClass) Default() {
+	// Default all namespaces in te references to the object namespace.
+	if len(in.Spec.Infrastructure.Ref.Namespace) == 0 {
+		in.Spec.Infrastructure.Ref.Namespace = in.Namespace
+	}
+	if len(in.Spec.ControlPlane.Ref.Namespace) == 0 {
+		in.Spec.ControlPlane.Ref.Namespace = in.Namespace
+	}
+	for i := range in.Spec.Worker.Deployments {
+		if len(in.Spec.Worker.Deployments[i].Template.Bootstrap.Ref.Namespace) == 0 {
+			in.Spec.Worker.Deployments[i].Template.Bootstrap.Ref.Namespace = in.Namespace
+		}
+		if len(in.Spec.Worker.Deployments[i].Template.Infrastructure.Ref.Namespace) == 0 {
+			in.Spec.Worker.Deployments[i].Template.Infrastructure.Ref.Namespace = in.Namespace
+		}
+	}
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateCreate() error {
+	return in.validate(nil)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateUpdate(old runtime.Object) error {
+	oldClusterClass, ok := old.(*ClusterClass)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a ClusterClass but got a %T", old))
+	}
+	return in.validate(oldClusterClass)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (in *ClusterClass) ValidateDelete() error {
+	return nil
+}
+
+func (in *ClusterClass) validate(old *ClusterClass) error {
+	var allErrs field.ErrorList
+
+	// ensure all the references are within the same namespace
```

Comment:
Have you considered using [`corev1.TypedLocalObjectReference`](https://github.com/kubernetes/api/blob/v0.21.2/core/v1/types.go#L5414) so it's just not possible to select a different namespace?

Reply:
TypedLocalObjectReference doesn't contain the Version of the APIs which is required in our case

Reply:
Resolving as per comment above

Reply:
I think its time to revive the **Long Lost** proposal I worked on last year 😆